src/ui/ToggleEditor.tsx

```diff
@@ -57,17 +90,47 @@ export default @CMBContext class ToggleEditor extends Component {
       parse: PropTypes.func.isRequired,
     }),
     options: PropTypes.object,
-    api: PropTypes.object,
+    api: PropTypes.shape({
+      setValue: PropTypes.func,
+      markText: PropTypes.func,
+      blockMode: PropTypes.bool,
+    }),
     appElement: PropTypes.instanceOf(Element).isRequired,
-    debuggingLog: PropTypes.object,
+    debuggingLog: PropTypes.shape({
+      history: PropTypes.any,
+    }),
   }
 
   static defaultProps = {
     debuggingLog: {},
     cmOptions: {},
   }
 
-  constructor(props) {
+  // TODO(pcardune): None of these should be here. Denormalizing
+  // props is a very bad thing to do.
+  cmOptions: ToggleEditorProps['cmOptions'] & typeof defaultCmOptions;
+  language: ToggleEditorProps['language'];
+  parse: ToggleEditorProps['language']['parse'];
+  getExceptionMessage: ToggleEditorProps['language']['getExceptionMessage'];
+  getASTNodeForPrimitive: ToggleEditorProps['language']['getASTNodeForPrimitive'];
+  getLiteralNodeForPrimitive: ToggleEditorProps['language']['getLiteralNodeForPrimitive'];
+  primitivesFn: ToggleEditorProps['language']['primitivesFn'];
+  options: ToggleEditorProps['options'];
+
+  // TODO(pcardune): remove this field, as we should be able to rely entirely
+  // on lifecycle methods without having to keep track of which life cycle
+  // methods have been called.
+  hasMounted: boolean;
+
+  currentCode: unknown;
+
+  eventHandlers: {};
+
+  toolbarRef: React.RefObject<Toolbar>;
+  ast: unknown;
+  newAST: AST;
```

Comment:
All of these fields living on the class instance give me pause and should be unnecessary in newer versions of react. But at least now they are typed....

Reply:
Wait, newer versions of react don't require this to be declared? Is that because of contexts (making better use of that feature has long been a wish of mine), or is there some magic I'm not aware of?

Reply:
There are a few different things going on.

**1. Fields that come from props.**

Instead of

```js
this.getASTNodeForPrimitive(...)
```

you should just use

```js
this.props.language.getASTNodeForPrimitive(...)
````

_unless_ those props are supposed to serve as initial values for component state, in which case you'd access them through the state object:

```js
this.state.language.getASTNodeForPrimitive(...)
```

By storing stuff on the component class, you are circumventing React's carefully crafted state management system, which makes it easier to introduce bugs and get unexpected behavior. For example, what if I have some UI outside of ToggleEditor where the user can choose which language they are using from a dropdown. The language they have chosen then gets passed down to the ToggleEditor component through props, and the ToggleEditor component should presumably re-render accordingly. But react isn't going to delete and reconstruct the ToggleEditor component because then any state in ToggleEditor (like whether blockMode is enabled or not) would get wiped out and reset. Instead react is going to reuse the existing instance of ToggleEditor and just update the props. But now the render() method won't be using the right language! It will use the old language that was passed when ToggleEditor was initially constructed and rendered. Now you have a bug.

React has always worked this way, and in the past if you really wanted to extract stuff from props into component instance variables, you'd have to make sure to re-copy all the stuff out of props again by implementing a [componentWillUpdate](https://reactjs.org/docs/react-component.html#unsafe_componentwillupdate) method on your component. But people kept screwing this up over the years, so now `componentWillUpdate` is deprecated. In fact, the use of class components instead of pure functional components is now considered outdated and bad practice, in part because it makes it so much easier to inadvertently mess up things like this.

**2. Sharing variables across multiple "sub-render" methods**

`currentCode` comes from `SHARED.cm.getValue()` and presumably refers to the text that the user has typed into the codemirror editor. `this.currentCode` later gets accessed in `renderBlocks` and `renderCode` and written in `handleToggle`, but it's difficult to reason about the intended semantics. Is `this.currentCode` only there so that it doesn't have to be explicitly passed to `renderBlocks` and `renderCode`? If the value coming from `SHARED.cm.getValue()`, which is effectively global state that can be changed from anywhere, were to be changed, should the component re-render? If so, then `renderBlocks` and `renderCode` will be using an old version and you have bug.

Once again you are denormalizing state management in a way that circumvents react, but this case is a bit different from (1.) because you are interfacing with a separate non-react library that does it's own state management. In the olden days, you would use `componentDidMount` and `componentWillUnmount` to set up whatever subscription handlers were necessary to subscribe to this outside state, and update your component's internal state whenever the subscription handlers were called, thereby synchronizing state across disparate systems. You would have to store the subscription object on the class component so it could be shared between `componentDidMount` and `componentWillUnmount` like so:

```js
componentDidMount() {
  this.subscription = SomeLibrary.onChange((newValue) => {
    this.setState({value: newValue});
  });
}
componentWillUnmount() {
  this.subscription.cancel();
}
```

This works so long as you are careful about not touching `this.subscription` outside those two methods. Nowadays though, this is no longer necessary and you're supposed to use [react effect hooks](https://reactjs.org/docs/hooks-effect.html), which lets you encapsulate this logic in a way that is both more reusable and doesn't require storing temporary variables on your component instance.

**3. Other stuff**

There's other stuff going on here too, but I need a break from writing this essay lol.

Reply:
This is extremely helpful - thanks for taking the time to explain it! It sounds like I should go through and take a crack re-organizing component state. If I put together a PR next week, would you be willing to review?

Reply:
Yep, I'd definitely be happy to review that. Aside from just cleaning up some of this component state, there is also the possibility of refactoring this code to the more modern functional style components that use react hooks. That's a bigger undertaking that would also force a cleanup of component state. But doing the smaller cleanup first will probably make it easier to do a bigger cleanup later.